### PR TITLE
[ENT-500] Add Program-based consent architecture

### PR DESCRIPTION
**Pull request by haikuginger**
**Description:**
This pull request adds the ability to retrieve and post consent information scoped for an entire program at once.

This pull request also allows for treatment of course keys as first-class citizens in the consent process in addition to the existing use of course run keys; if a record with a course key is stored, it will answer to requests for any course run that's an instance of that course.

The programs consent feature takes advantage of this in that programs don't necessarily have canonical course runs; therefore, rather than consenting to any particular course run, the user is, behind the scenes, consenting to all the courses in the program.

**JIRA:** [ENT-500](https://openedx.atlassian.net/browse/ENT-500)

**Dependencies:** None

**Merge deadline:** 28 August 2017

**Installation requirements:**
- Course discovery set up
  - Program with courses
  - Catalog that includes all courses in the program
- EnterpriseCustomer set up
  - Catalog set to aforementioned catalog
  - Data sharing consent enabled/set to enforce at enrollment

**Testing instructions (API):**

1. Browse to URL matching pattern:
    ```
    /consent/api/v1/data_sharing_consent?enterprise_customer_uuid=your-ec-uuid-here&username=username&program_uuid=your-program-uuid-here
    ```
2. In the LMS shell, instantiate an API client with a staff user, aimed at `/consent/api/v1`.
3. Test the following with your client, verifying the expected behavior, and noting that calling the course-specific API with both course run IDs and the overall course IDs yields the expected result.:
    ```python
    # Start by verifying that no records appear to exist
    >>> client.data_sharing_consent.get(enterprise_customer_uuid='ec-uuid', username='username', program_uuid='program_uuid')
    {'enterprise_customer_uuid': 'ec-uuid', 'username': 'username', 'program_uuid': 'program_uuid', 'exists': False, 'consent_provided': False, 'consent_required': True}

    >>> client.data_sharing_consent.get(enterprise_customer_uuid='ec-uuid', username='username', course_id='course-v1:org+course+specific_course_run')
    {'enterprise_customer_uuid': 'ec-uuid', 'username': 'username', 'course_id': 'course-v1:org+course+specific_course_run', 'exists': False, 'consent_provided': False, 'consent_required': True}

    >>> client.data_sharing_consent.get(enterprise_customer_uuid='ec-uuid', username='username', course_id='org+course')
    {'enterprise_customer_uuid': 'ec-uuid', 'username': 'username', 'course_id': 'org+course', 'exists': False, 'consent_provided': False, 'consent_required': True}

    # Provide program-scoped consent
    >>> client.data_sharing_consent.post({'enterprise_customer_uuid': 'ec-uuid', 'username': 'username', 'program_uuid': 'program_uuid'})
    {'enterprise_customer_uuid': 'ec-uuid', 'username': 'username', 'program_uuid': 'program_uuid', 'exists': True, 'consent_provided': True, 'consent_required': False}

    # Verify that individual course run consent now has been provided
    >>> client.data_sharing_consent.get(enterprise_customer_uuid='ec-uuid', username='username', course_id='course-v1:org+course+specific_course_run')
    {'enterprise_customer_uuid': 'ec-uuid', 'username': 'username', 'course_id': 'course-v1:org+course+specific_course_run', 'exists': True, 'consent_provided': True, 'consent_required': False}

    # As well as for the course
    >>> client.data_sharing_consent.get(enterprise_customer_uuid='ec-uuid', username='username', course_id='org+course')
    {'enterprise_customer_uuid': 'ec-uuid', 'username': 'username', 'course_id': 'org+course', 'exists': True, 'consent_provided': True, 'consent_required': False}

    # Delete consent for all the items in the program
    >>> client.data_sharing_consent.delete(enterprise_customer_uuid='ec-uuid', username='username', program_uuid='program_uuid')
    {'enterprise_customer_uuid': 'ec-uuid', 'username': 'username', 'program_uuid': 'program_uuid', 'exists': True, 'consent_provided': False, 'consent_required': True}

    # Validate revocation on the course/course run
    >>> client.data_sharing_consent.get(enterprise_customer_uuid='ec-uuid', username='username', course_id='course-v1:org+course+specific_course_run')
    {'enterprise_customer_uuid': 'ec-uuid', 'username': 'username', 'course_id': 'course-v1:org+course+specific_course_run', 'exists': True, 'consent_provided': False, 'consent_required': True}

    >>> client.data_sharing_consent.get(enterprise_customer_uuid='ec-uuid', username='username', course_id='org+course')
    {'enterprise_customer_uuid': 'ec-uuid', 'username': 'username', 'course_id': 'org+course', 'exists': True, 'consent_provided': False, 'consent_required': True}
    ```

**Testing instructions (UI):**

1. Delete any consent records created in the API testing.
2. Create a new (non-staff) user account.
3. Log into that user.
4. Browse to the URL matching the following pattern, replacing the relevant UUIDs with your own (copy it to an external note; you'll need it again):
    ```
    http://localhost:18000/enterprise/grant_data_sharing_permissions?enterprise_customer_uuid=8019d1dc-1116-4e14-8629-2e5306c13c88&program_uuid=0594a613-08e4-45e7-92ca-39b488e46f34&next=https%3A%2F%2Fgoogle.com%2F&&failure_url=https%3A%2F%2Fnews.ycombinator.com%2F
    ```
5. Note that you see a consent prompt.
6. Decline consent.
7. Note that you're redirected to HackerNews (required parameter in the URL, and we don't have other handlers yet).
8. Browse to the same URL again.
9. This time, provide consent.
10. Note that you're redirected to Google (as above).
11. Attempt to browse to the URL again and note that you get a 404 because consent is no longer needed.
12. Delete the resulting consent record(s) from the database.
13. Repeat steps 4 through 10, adding an extra url parameter: `&enrollment_deferred=True`.
14. Repeat step 11, but note that the page is loadable again; the `enrollment_deferred` parameter resulted in the record not being committed to the database, which means it's still required.

**Merge checklist:**

- [x] Check that the versions of the requirements in the `platform-master.in` file match edx-platform.
- [x] New requirements are in the right place (`base.in` if only used in enterprise; in the correct `platform-****.in` files if they're hosted in edx-platform)
- [x] Regenerate requirements with `make upgrade && make requirements` (and make sure to fix any errors).
  **DO NOT** just add dependencies to `requirements/*.txt` files.
- [x] All reviewers approved
- [x] CI build is green
- [x] Version bumped
- [x] Changelog record added
- [x] Documentation updated (not only docstrings)
- [x] Commits are (reasonably) squashed
- [x] Translations are updated
- [x] PR author is listed in AUTHORS

**Post merge:**
- [ ] Create a tag
- [ ] Check new version is pushed to PyPi after tag-triggered build is finished.
- [ ] Delete working branch (if not needed anymore)
- [ ] edx-platform PR (be sure to include edx-platform requirements upgrades that were present in this PR)

### CODE REVIEW

**Comment by edx-webhook** [comment]
Thanks for the pull request, @haikuginger! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request.

<a   href="https://openedx-webhooks.herokuapp.com/github/process_pr?number=173&repo=edx%2Fedx-enterprise"   rel="nofollow" >Create an OSPR issue for this pull request</a>.

**Comment by haikuginger** [comment]
@brittneyexline @itsjeyd, this is on the business sandbox for testing.

File: consent/models.py [hunk 1]
```diff
@@ -26,17 +28,35 @@ class DataSharingConsentQuerySet(models.query.QuerySet):
     except is not saved in the database until committed.
     """
 
-    def get(self, *args, **kwargs):
+    def proxied_get(self, *args, **kwargs):
```
Comment (brittneyexline) [inline consent/models.py]: nit: i'm not sure about this name - it's not intuitively clear what it means, but i don't have any alternatives to suggest.

File: consent/api/v1/views.py [hunk 2]
```diff
@@ -129,19 +146,43 @@ def get_required_query_params(self, request):
...
+            self.CONSENT_GRANTED: False,
+            self.CONSENT_REQUIRED: False,
+        }
+        if course_id:
+            data[self.REQUIRED_PARAM_COURSE_ID] = course_id
+
+        if program_id:
+            data[self.REQUIRED_PARAM_PROGRAM_ID] = program_id
+
+        return Response(data, status=HTTP_200_OK)
```
Comment (haikuginger) [inline consent/api/v1/views.py]: @douglashall requested that we switch to a 200 from the previous 404 we were throwing for nonexistent records.

**Comment by UmanShahzad** [comment]
@haikuginger I'll be reviewing this soon -- but I'm a *little* worried about starting the review while the huge number of conflicts above exist... o_O What's your call on this? Should I give it a go or wait on conflict resolution?

File: consent/models.py [hunk 3]
```diff
@@ -26,17 +28,35 @@ class DataSharingConsentQuerySet(models.query.QuerySet):
...
+                # to the second check.
+                pass
+            else:
+                try:
+                    # Try to get the record for the course run specifically
+                    return self.get(*args, **kwargs)
+                except DataSharingConsent.DoesNotExist:
+                    # A record for the course run didn't exist, so modify the query
+                    # parameters to look for just a course record on the second pass.
+                    kwargs['course_id'] = course_run_key.org + '+' + course_run_key.course
```
Comment (haikuginger) [inline consent/models.py]: Yes; making that change.

File: consent/models.py (hunk 1 above)
Comment (haikuginger) [inline consent/models.py]: Yeah, I'd be open to suggestions, but I don't have anything better off the top of my head.

**Comment by haikuginger** [comment]
Rebased and nit addressed from 549e0e5. @UmanShahzad, this should be relatively safe to review now.

**Comment by UmanShahzad** [comment]
@haikuginger Thanks! Will be getting to it.

File: tests/test_enterprise/api_client/test_discovery.py [hunk 4]
```diff
@@ -401,12 +401,23 @@ def get_course_run(*args, **kwargs):  # pylint: disable=unused-argument
...
     @ddt.unpack
-    def test_is_course_in_catalog(self, catalog_id, course_id, api_resp, expected, mock_discovery_client_factory):
+    def test_is_course_in_catalog(
+            self,
+            catalog_id,
+            course_id,
+            api_resp,
+            expected,
+            is_run,
+            mock_discovery_client_factory
```
Comment (UmanShahzad) [inline tests/test_enterprise/api_client/test_discovery.py]: Can we put the mock patch next to the function name? I notice we usually do that as it's easier to catch. For a moment I was wondering how in the world those booleans could be a `mock_discovery_client_factory`, but then I looked up.

File: tests/test_enterprise/api_client/test_discovery.py [hunk 5]
```diff
@@ -401,12 +401,23 @@ def get_course_run(*args, **kwargs):  # pylint: disable=unused-argument
...
     )
     @ddt.unpack
-    def test_is_course_in_catalog(self, catalog_id, course_id, api_resp, expected, mock_discovery_client_factory):
+    def test_is_course_in_catalog(
+            self,
+            catalog_id,
+            course_id,
+            api_resp,
+            expected,
+            is_run,
```
Comment (UmanShahzad) [inline tests/test_enterprise/api_client/test_discovery.py]: Nit: `is_a_course_run`

File: tests/test_enterprise/api_client/test_discovery.py [hunk 6]
```diff
@@ -416,7 +427,21 @@ def test_is_course_in_catalog(self, catalog_id, course_id, api_resp, expected, m
...
+            discovery_client.catalogs.return_value.contains.get.assert_called_once_with(course_run_id=course_id)
+        else:
+            discovery_client.catalogs.return_value.contains.get.assert_called_once_with(course_id=course_id)
+
+    @ddt.data(
+        (None, []),
+        ({}, []),
+        ({'courses': [{'key': 'first+key'}, {'key': 'second+key'}]}, ['first+key', 'second+key'])
+    )
+    @ddt.unpack
```
Comment (UmanShahzad) [inline tests/test_enterprise/api_client/test_discovery.py]: Nit: at least for the really nested stuff, can we unnest them?

Comment (UmanShahzad) [inline tests/test_enterprise/api_client/test_discovery.py]: BTW by 'unnest' I meant put them on separate lines -- sorry if that was confusing.

File: consent/models.py [hunk 7]
```diff
@@ -81,26 +112,63 @@ def __init__(self, enterprise_customer=None, username='', course_id='', granted=
...
+        if self._child_consents:
+            consents = []
+
+            for consent in self._child_consents:
+                consent.granted = self.granted
+                consents.append(consent.save() or consent)
+
+            return ProxyDataSharingConsent.from_children(self.program_id, *consents)
+
+        consent, _ = DataSharingConsent.objects.update_or_create(
```
Comment (haikuginger) [inline consent/models.py]: Unless we're in a view where we use `_` as shorthand for `ugettext`.

File: tests/test_enterprise/api_client/test_discovery.py (hunk 6 above)
Comment (haikuginger) [inline tests/test_enterprise/api_client/test_discovery.py]: For nesting in the DDT cases, I'd rather leave them as-is. While it can look a bit hard to parse, I think it's easier to accurately convey the components things we're doing, and we're less likely to screw up recomposition of individual arguments. For this specific case and one other, I've re-whitespaced to make it a bit easier to parse.

File: consent/models.py (hunk 7 above)
Comment (UmanShahzad) [inline consent/models.py]: Ah!

File: tests/test_enterprise/api_client/test_discovery.py (hunk 6 above)
Comment (UmanShahzad) [inline tests/test_enterprise/api_client/test_discovery.py]: Sounds good to me :+1: And thanks!

File: consent/helpers.py [hunk 8]
```diff
@@ -71,10 +81,18 @@ def get_course_data_sharing_consent(username, course_id, enterprise_customer_uui
         enterprise_customer__uuid=enterprise_customer_uuid
     )
 
+
 def get_program_data_sharing_consent(username, program_id, enterprise_customer_uuid):
+    """
+    Get the data sharing consent object associated with a certain user of a customer for a program.
+
+    :param username: The user that grants consent.
+    :param program_id The program for which consent is granted.
```
Comment (UmanShahzad) [inline consent/helpers.py]: Nit: missing ":" at the end of program_id. With that said, maybe rename to `program_uuid` everywhere as we changed what it's also referred to as in the view? If that's too much work, nevermind.

**Comment by haikuginger** [comment]
@UmanShahzad, rebased on master and re-targeted PR, plus some squashing to make segmentation of changes more obvious.

**Comment by UmanShahzad** [comment]
@haikuginger I tested this, following your instructions. To get a client, I did the following (docker devstack):

```sh
>>> from django.contrib.auth.models import User
>>> user = User.objects.create(username='staffx', is_staff=True)
>>> from edx_rest_api_client.client import EdxRestApiClient
>>> from openedx.core.lib.token_utils import JwtBuilder
>>> client = EdxRestApiClient('http://localhost:18000/consent/api/v1/', append_slash=False, jwt=JwtBuilder(user).build_token(['profile', 'email'], 99999))

>>> client.data_sharing_consent.get(enterprise_customer_uuid='1c4140b9-e149-4170-893d-7dec9aa39c29', username='lowlyuser', program_uuid='0fd321b8-3f15-4541-8224-3cf081923e34')
{u'username': u'lowlyuser', u'enterprise_customer_uuid': u'1c4140b9-e149-4170-893d-7dec9aa39c29', u'exists': False, u'program_uuid': u'0fd321b8-3f15-4541-8224-3cf081923e34', u'consent_provided': False, u'consent_required': True}

>>> client.data_sharing_consent.post({'enterprise_customer_uuid':'1c4140b9-e149-4170-893d-7dec9aa39c29', 'username':'lowlyuser', 'program_uuid':'0fd321b8-3f15-4541-8224-3cf081923e34'})
{u'username': u'lowlyuser', u'enterprise_customer_uuid': u'1c4140b9-e149-4170-893d-7dec9aa39c29', u'exists': True, u'program_uuid': u'0fd321b8-3f15-4541-8224-3cf081923e34', u'consent_provided': True, u'consent_required': False}

>>> client.data_sharing_consent.delete(enterprise_customer_uuid='1c4140b9-e149-4170-893d-7dec9aa39c29', username='lowlyuser', program_uuid='0fd321b8-3f15-4541-8224-3cf081923e34')
True

>>> client.data_sharing_consent.get(enterprise_customer_uuid='1c4140b9-e149-4170-893d-7dec9aa39c29', username='lowlyuser', course_id='course-v1:edX+DemoX+Demo_Course')
{u'username': u'lowlyuser', u'enterprise_customer_uuid': u'1c4140b9-e149-4170-893d-7dec9aa39c29', u'exists': True, u'course_id': u'edX+DemoX', u'consent_provided': False, u'consent_required': True}

>>> client.data_sharing_consent.post(enterprise_customer_uuid='1c4140b9-e149-4170-893d-7dec9aa39c29', username='lowlyuser', course_id='course-v1:edX+DemoX+Demo_Course')
{u'username': u'lowlyuser', u'enterprise_customer_uuid': u'1c4140b9-e149-4170-893d-7dec9aa39c29', u'exists': True, u'course_id': u'edX+DemoX', u'consent_provided': True, u'consent_required': False}

>>> client.data_sharing_consent.delete(enterprise_customer_uuid='1c4140b9-e149-4170-893d-7dec9aa39c29', username='lowlyuser', course_id='course-v1:edX+DemoX+Demo_Course')
True

>>> client.data_sharing_consent.get(enterprise_customer_uuid='1c4140b9-e149-4170-893d-7dec9aa39c29', username='lowlyuser', course_id='edX+DemoX')
{u'username': u'lowlyuser', u'enterprise_customer_uuid': u'1c4140b9-e149-4170-893d-7dec9aa39c29', u'exists': True, u'course_id': u'edX+DemoX', u'consent_provided': False, u'consent_required': True}

>>> client.data_sharing_consent.post(enterprise_customer_uuid='1c4140b9-e149-4170-893d-7dec9aa39c29', username='lowlyuser', course_id='edX+DemoX')
{u'username': u'lowlyuser', u'enterprise_customer_uuid': u'1c4140b9-e149-4170-893d-7dec9aa39c29', u'exists': True, u'course_id': u'edX+DemoX', u'consent_provided': True, u'consent_required': False}

>>> client.data_sharing_consent.delete(enterprise_customer_uuid='1c4140b9-e149-4170-893d-7dec9aa39c29', username='lowlyuser', course_id='edX+DemoX')
True
```

For some reason, `DELETE` is just returning `True`.

File: consent/api/v1/views.py [hunk 9]
```diff
@@ -69,6 +69,7 @@ class DataSharingConsentView(APIView):
 
     REQUIRED_PARAM_USERNAME = 'username'
     REQUIRED_PARAM_COURSE_ID = 'course_id'
+    REQUIRED_PARAM_PROGRAM_UUID = 'program_uuid'
```
[the remark this answers is not in the document]
Comment (haikuginger) [reply consent/api/v1/views.py]: Good catch.

**Comment by UmanShahzad** [comment]
I think `DELETE` may have been returning just `True` because I wasn't passing in a dictionary? Let me try.

**Comment by haikuginger** [comment]
@UmanShahzad I think the CodeCov actually isn't accurate for some reason; possibly because I pushed changes before switching the base branch. As far as the return value of the DELETE call, it looks like that's an artifact of [how Slumber](https://github.com/samgiles/slumber/blob/master/slumber/__init__.py#L175-L183) processes DELETE requests. Will address other comments, add testing for the Program user-facing views, and add testing instructions for the user-facing views.

**Comment by UmanShahzad** [comment]
@haikuginger Ah ok, and nevermind about trying DELETE with a dictionary... I think you're right about Slumber.

**Comment by haikuginger** [comment]
@UmanShahzad, I've updated with GET tests for the program DSC view, as well as updated error messaging from the API.

**Comment by UmanShahzad** [comment]
@haikuginger Thanks so much for the renames from `x_id` -> `x_uuid` :)

**Comment by haikuginger** [comment]
@UmanShahzad, testing instructions in the description. Going to finish up the POST tests later.

**Comment by UmanShahzad** [comment]
@haikuginger I will give this a test in a moment.

**Comment by UmanShahzad** [comment]
@haikuginger Btw I just wanted to note, but not sure how important this is: course-specific consent expects `enterprise_id` as a query param, whereas program-specific consent expects `enterprise_customer_uuid`. I don't think that's a problem in practice, but it can cause some confusion, as it did for me when trying to figure out which to use initially for my own code.

**Comment by haikuginger** [comment]
@UmanShahzad, I agree it can be confusing. The course-scoped consent view right now has a lot of nasty implicit stuff and archaic naming. I wanted to implement the Program view with explicit items and updated naming for us to use as an example when we go to update the course-scoped view.

**Comment by UmanShahzad** [comment]
@haikuginger Good plan, and good luck!

**Comment by UmanShahzad** [comment]
@haikuginger I just remembered to say this: it turns out there's some coverage missing from the view. I noticed when I rebased #171 on this.

**Comment by haikuginger** [comment]
@UmanShahzad, I'm pretty sure that's spurious; running tests locally shows 100% coverage on everything.

**Comment by UmanShahzad** [comment]
@haikuginger Hmmm, that's interesting. Here's what I get on doing this on `uman/program-landing-page`:

```
enterprise/views.py                                                                        423     21    104      1    94%   461-485, 501-504, 499->501
```

And on `haikuginger/program-dsc-api`:

```
enterprise/views.py                                                                        318     21     80      1    92%   442-466, 482-485, 480->482
```

Line differ only due to my additions to imports etc..

Can you double-check it? I even ran `make clean` which calls `coverage erase`.

**Comment by haikuginger** [comment]
@UmanShahzad, Whoops, you're right - I still don't have POST coverage. I'm tempted to close and re-open this as a separate PR once I have that to make sure Codecov is accurate.

Sorry, I was thinking of a different PR's test results.

File: enterprise/views.py [hunk 10]
```diff
@@ -291,23 +291,95 @@ def get_course_specific_consent(self, request, course_id):
...
-                    _('your enrollment in all sponsored courses'),
-                    _('your learning progress'),
-                    _('course completion'),
-                ],
-            })
+
+        return render(request, 'enterprise/grant_data_sharing_permissions.html', context=context_data)
+
+    @method_decorator(login_required)
+    def get_program_specific_consent(self, request, program_uuid):
```
Comment (brittneyexline) [inline enterprise/views.py]: it looks like this has a lot in common with the course specific function, is there more opportunity to consolidate, particularly with the context? although based on the conversation in this pr it sounds like something you plan to address later when cleaning up the course specific logic.

Comment (haikuginger) [reply enterprise/views.py]: Yeah, cleanup/consolidation will come later.

**Comment by UmanShahzad** [comment]
@haikuginger Tests look good to me! I think you're good to squash and merge once the tests & coverage pass.

**Comment by UmanShahzad** [comment]
(oh and bump version, etc.)

**Comment by haikuginger** [comment]
Rebased, squashed, and final touches from a223e6c.
